ClipVessel/ClipVessel.py

```diff
@@ -0,0 +1,800 @@
+  """
+  def __init__(self):
+    ScriptedLoadableModuleLogic.__init__(self)
+    self.radiusArrayName = 'Radius'
+    self.blankingArrayName = 'Blanking'
+    self.groupIdsArrayName = 'GroupIds'
+    self.tractIdsArrayName = 'TractIds'
+    self.centerlineIdsArrayName = 'CenterlineIds'
+    
+    self.gapLength = 1.0
+    self.tolerance = 0.01
+    self.clipValue = 0.0
+    self.cutoffRadiusFactor = 1E16
+
+    self.groupIds = []
+
+    self.useRadiusInformation = 1
+
+    self.Sigma = 1
+    self.AdaptiveExtensionLength = 0
+    self.AdaptiveExtensionRadius = 1
+    self.AdaptiveNumberOfBoundaryPoints = 0
+    self.ExtensionRatio = 2
+    self.ExtensionRadius = 1
+    self.TransitionRatio = 0.25
+    self.CenterlineNormalEstimationDistanceRatio = 1.0
+    self.TargetNumberOfBoundaryPoints = 50
+    
+  def setDefaultParameters(self, parameterNode):
+    """
+    Initialize parameter node with default settings.
+    """
+    # We choose a small target point number value, so that we can get fast speed
+    # for smooth meshes. Actual mesh size will mainly determined by DecimationAggressiveness value.
+    if not parameterNode.GetParameter("TargetNumberOfPoints"):
+        parameterNode.SetParameter("TargetNumberOfPoints", "50000")
+    if not parameterNode.GetParameter("DecimationAggressiveness"):
+        parameterNode.SetParameter("DecimationAggressiveness", "4.0")
+    if not parameterNode.GetParameter("PreprocessInputSurface"):
+        parameterNode.SetParameter("PreprocessInputSurface", "true")
+    if not parameterNode.GetParameter("SubdivideInputSurface"):
+        parameterNode.SetParameter("SubdivideInputSurface", "false")
+    if not parameterNode.GetParameter("ExtensionLength"):
+        parameterNode.SetParameter("ExtensionLength", "5")
+        
+  def polyDataFromNode(self, surfaceNode, segmentId):
+    if not surfaceNode:
+        logging.error("Invalid input surface node")
+        return None
+    if surfaceNode.IsA("vtkMRMLModelNode"):
+        return surfaceNode.GetPolyData()
+    elif surfaceNode.IsA("vtkMRMLSegmentationNode"):
+        # Segmentation node
+        polyData = vtk.vtkPolyData()
+        surfaceNode.CreateClosedSurfaceRepresentation()
+        surfaceNode.GetClosedSurfaceRepresentation(segmentId, polyData)
+        return polyData
+    else:
+        logging.error("Surface can only be loaded from model or segmentation node")
+        return None
+            
+  def capSurface(self, surface):
+    capDisplacement = 0.0
+    surfaceCapper = vtkvmtkComputationalGeometry.vtkvmtkCapPolyData()
+    surfaceCapper.SetInputData(surface)
+    surfaceCapper.SetDisplacement(capDisplacement)
+    surfaceCapper.SetInPlaneDisplacement(capDisplacement)
+    surfaceCapper.Update()
+    surface = surfaceCapper.GetOutput()
+    return surface
+    
+  def preprocess(self, surfacePolyData, targetNumberOfPoints, decimationAggressiveness, subdivide):
```

Comment:
Duplicating code would significantly increase maintenance workload. Please call preprocess method of ExtractCenterline.py instead. You can do something like this:

```
import ExtractCenterline
extractCenterlineLogic = ExtractCenterline.ExtractCenterlineLogic()
extractCenterlineLogic.preprocess(...)
```